src/ucai/core/utils/callable_utils.py

```diff
@@ -0,0 +1,361 @@
+    dedents the code, and returns the indentation unit used in the function (e.g., 2 or 4 spaces).
+    """
+    source_lines, _ = inspect.getsourcelines(func)
+    dedented_source = dedent(''.join(source_lines))
+    func_name = func.__name__
+
+    extractor = FunctionBodyExtractor(func_name, dedented_source)
+    parsed_source = ast.parse(dedented_source)
+    extractor.visit(parsed_source)
+
+    return extractor.function_body, extractor.indent_unit
+
+def validate_type_hint(hint: Any) -> str:
+    """Validates and returns the SQL type for a given type hint."""
+    # Handle typing.Optional, which is Union[type, None]
+    if hasattr(hint, '__origin__') and hint.__origin__ is Union:
+        non_none_types = [t for t in hint.__args__ if t is not type(None)]
+        if len(non_none_types) == 1:
+            return python_type_to_sql_type(non_none_types[0])
+        else:
+            raise ValueError(f"Unsupported type union: {hint}")
+    if hint is Any:
+        raise ValueError("Unsupported Python type: typing.Any is not allowed. Please specify a concrete type.")
+    return python_type_to_sql_type(hint)
+
+def generate_type_hint_error_message(param_name: str, param_hint: Any, exception: Exception) -> str:
+    """
+    Generate an informative error message for unsupported parameter types, especially for Lists, Tuples, and Dicts.
+    Args:
+        param_name: The name of the parameter with the type hint issue.
+        param_hint: The unsupported type hint.
+        exception: The original exception raised.
+    Returns:
+        str: A detailed error message guiding the user on how to resolve the issue.
+    """
+    if hasattr(param_hint, '__origin__'):
+        origin = param_hint.__origin__
+        if origin == list:
+            return (
+                f"Error in parameter '{param_name}': List type requires a specific element type. "
+                f"Please define the internal type for the list, e.g., List[int]. Original error: {exception}"
+            )
+        elif origin == tuple:
+            return (
+                f"Error in parameter '{param_name}': Tuple type requires specific element types. "
+                f"Please define the types for the tuple elements, e.g., Tuple[int, str]. Original error: {exception}"
+            )
+        elif origin == dict:
+            return (
+                f"Error in parameter '{param_name}': Dict type requires both key and value types. "
+                f"Please define the internal types for the dict, e.g., Dict[str, int]. Original error: {exception}"
+            )
+    return (
+        f"Error in parameter '{param_name}': type {param_hint} is not supported. "
+        f"Original error: {exception}"
+    )
+
+def format_default_value(default: Any) -> str:
+    """Formats the default value for SQL."""
+    if default is None:
+        return 'NULL'
+    elif isinstance(default, str):
+        return f"'{default}'"
+    else:
+        return str(default)
```

Comment:
Should we forbidden array and map default values?

Reply:
Yeah good call. Defaults in collections are always odd. Added validation and tests to confirm